Src/HALALMock/Services/Time/Time.cpp

```diff
@@ -1,445 +1,168 @@
-        low_precision_ids++;
-    uint8_t id = low_precision_ids;
-    uint64_t tick_on_register = low_precision_tick;
-    Time::register_low_precision_alarm(
-        milliseconds, [&, id, callback, tick_on_register]() {
-            if (tick_on_register == low_precision_tick) {
-                return;
-            }
-            callback();
-            Time::unregister_low_precision_alarm(id);
-        });
-    return id;
-}
+        if (real_wait_duration <= RealClock::duration::zero()) {
+            // Real time cannot keep up with simulation time
+            //TODO: ErrorHandler("Real time cannot keep up with simulation time.");
+            break;
+        }
 
-void Time::cancel_timeout(uint8_t id) {
-    // we can take advantage of the fact that if we try to
-    // unregister an alarm that is not registered, it will return false
-    // instead of going HardFault or ErrorHandler
-    Time::unregister_low_precision_alarm(id);
-}
+        //Go to sleep until next alarm
+        cv.wait_for(lock, real_wait_duration, [] { return !running; });
 
-void Time::global_timer_callback() {
-    Time::global_tick += Time::HIGH_PRECISION_MAX_ARR;
-}
+        if (!running) break;
 
-void Time::high_precision_timer_callback(TIM_HandleTypeDef* tim) {
-    Time::high_precision_alarms_by_timer[tim].alarm();
-}
+        now_real = RealClock::now();
+        simulation_time = std::chrono::duration_cast<SimDuration>((now_real - simulation_start_time) * simulation_speed);
 
-void Time::mid_precision_timer_callback() {
-    for (pair<const uint8_t, Time::Alarm>& pair :
-         Time::mid_precision_alarms_by_id) {
-        Time::Alarm& alarm = pair.second;
-        if (alarm.is_on == false || alarm.offset == Time::mid_precision_tick)
-            continue;
-        if ((Time::mid_precision_tick - alarm.offset) % alarm.period == 0) {
-            alarm.alarm();
+        // Check if real time has fallen behind simulation time significantly
+        auto expected_simulation_time = std::chrono::duration_cast<SimDuration>((now_real - simulation_start_time) * simulation_speed);
+        lag = expected_simulation_time - simulation_time;
+        if (lag > max_lag) {
+            //ErrorHandler("Simulation time is significantly ahead of real time.");
+            break;
         }
-    }
-    Time::mid_precision_tick++;
-    while (not Time::mid_precision_erasable_ids.empty()) {
-        Time::mid_precision_alarms_by_id.erase(
-            Time::mid_precision_erasable_ids.top());
-        Time::mid_precision_erasable_ids.pop();
-    }
-}
 
-void Time::low_precision_timer_callback() {
-    for (auto& pair : Time::low_precision_alarms_by_id) {
-        Time::Alarm& alarm = pair.second;
-        if (alarm.is_on == false || alarm.offset == Time::low_precision_tick)
-            continue;
-        if ((Time::low_precision_tick - alarm.offset) % alarm.period == 0) {
-            alarm.alarm();
+        for (auto it = alarms.begin(); it != alarms.end();) {
```

Comment:
nitpick use structure binding for clearer naming

Reply:
what do you mean?

Reply:
```c++
for( auto& [id, alarm] : alarms){
    if(alarm.active && alarm.next_trigger <= simulation_time){
       // more code here
    }
}
```

Reply:
cant do that, the reason i am doing it with iterators is because i am deleting items as i iterate through the map, doing that like that there is no way of skipping an item when erased

Reply:
Completely right, I though you could get away with map::erase(const Key& key) but it segfault